ext/standard/uuid.c

```diff
@@ -0,0 +1,646 @@
+}
+
+static zend_always_inline zend_object *throw_argument_count_error(const char *method, const size_t passed, const size_t expected)
+{
+	return zend_throw_exception_ex(
+		zend_ce_argument_count_error,
+		0,
+		"Too few arguments to method UUID::%s(), %u passed and exactly %u expected",
+		method,
+		passed,
+		expected
+	);
+}
+
+PHP_METHOD(UUID, __construct) { /* NOOP */ }
+
+PHP_METHOD(UUID, fromBinary)
+{
+	char *input = NULL;
+	size_t input_len;
+
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, MIN(1, ZEND_NUM_ARGS()), "s", &input, &input_len) == FAILURE) {
+		throw_argument_count_error("fromBinary", ZEND_NUM_ARGS(), 1);
+		return;
+	}
+
+	if (input_len != PHP_UUID_LEN) {
+		zend_throw_exception_ex(
+			spl_ce_InvalidArgumentException,
+			0,
+			"Expected exactly %u bytes, but got %u",
+			PHP_UUID_LEN,
+			input_len
+		);
+		return;
+	}
+
+	new_uuid(return_value, input);
+}
+
+PHP_METHOD(UUID, parse)
+{
+	char *input = NULL;
+	size_t input_len;
+	php_uuid uuid;
+
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, MIN(ZEND_NUM_ARGS(), 1), "s", &input, &input_len) == FAILURE) {
+		throw_argument_count_error("parse", ZEND_NUM_ARGS(), 1);
+		return;
+	}
+
+	php_uuid_parse_throw(&uuid, input, input_len);
+	new_uuid(return_value, uuid);
+}
+
+PHP_METHOD(UUID, v3)
+{
+	zval *namespace = NULL;
+	char *name      = NULL;
+	size_t name_len;
+	php_uuid uuid;
+
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, MIN(ZEND_NUM_ARGS(), 2), "zs", &namespace, &name, &name_len) == FAILURE) {
+		throw_argument_count_error("v3", ZEND_NUM_ARGS(), 2);
+		return;
+	}
+
+	php_uuid_create_v3(&uuid, get_bytes(namespace), name, name_len);
+	new_uuid(return_value, uuid);
+}
+
+PHP_METHOD(UUID, v4)
+{
+	php_uuid uuid;
+	if (php_uuid_create_v4_throw(&uuid) == SUCCESS) {
+		new_uuid(return_value, uuid);
+	}
+}
+
+PHP_METHOD(UUID, v5)
+{
+	zval *namespace = NULL;
+	char *name      = NULL;
+	size_t name_len;
+	php_uuid uuid;
+
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, MIN(ZEND_NUM_ARGS(), 2), "zs", &namespace, &name, &name_len) == FAILURE) {
+		throw_argument_count_error("v5", ZEND_NUM_ARGS(), 2);
+		return;
+	}
+
+	php_uuid_create_v5(&uuid, get_bytes(namespace), name, name_len);
+	new_uuid(return_value, uuid);
+}
+
+PHP_METHOD(UUID, NamespaceDNS)  { new_uuid(return_value, PHP_UUID_NAMESPACE_DNS);  }
+PHP_METHOD(UUID, NamespaceOID)  { new_uuid(return_value, PHP_UUID_NAMESPACE_OID);  }
+PHP_METHOD(UUID, NamespaceURL)  { new_uuid(return_value, PHP_UUID_NAMESPACE_URL);  }
+PHP_METHOD(UUID, NamespaceX500) { new_uuid(return_value, PHP_UUID_NAMESPACE_X500); }
+PHP_METHOD(UUID, Nil)           { new_uuid(return_value, PHP_UUID_NIL);            }
+
+PHP_METHOD(UUID, __clone) { zend_throw_error(zend_ce_error, "Cannot clone immutable UUID object");                     }
+PHP_METHOD(UUID, __set)   { zend_throw_error(zend_ce_error, "Cannot set dynamic properties on immutable UUID object"); }
+
+PHP_METHOD(UUID, __wakeup)
+{
```

Comment:
This is like a punch in the face of everyone. Now every caller needs to check for `NULL` all over the place, just because we are not able to create a type-safe API due to concerns that our internal code works the same way as user code. Sorry, this makes no sense.